### ISSUE TITLE
crypto/ecdsa: Verify signature after signing to prevent faults from leaking private key

### ISSUE THREAD

**Issue by aarongable** [golang/go#54681]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.19 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes.

### What did you do?

It has been well known for [quite some time](https://www.researchgate.net/publication/2247203_On_the_Importance_of_Checking_Cryptographic_Protocols_for_Faults) that a single signature fault (e.g. due to a bit-flip or truncation) can lead to leaking the private key producing that signature. The crypto/rsa package has [contained a defense](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/rsa/rsa.go;l=632-637;drc=0a52d80666ddaa557cec17ad9166e2514b0bb6d4) against just such circumstances [since 2015](https://go-review.googlesource.com/c/go/+/17862/).

However, it is also possible for signature faults to leak ECDSA private keys. In particular, it [has been shown](https://eprint.iacr.org/2017/1014) that ECDSA private keys can be leaked if the same message is signed twice and one of those two signatures suffers a fault. This attack was [recently demonstrated](https://www.usenix.org/system/files/sec22-sullivan.pdf) (PDF link) to be practical.

### What did you expect to see?

I expected that ECDSA signatures would be validated immediately after being produced, in order to prevent faulty signatures from being exposed to an outside attacker.

### What did you see instead?

I do not see any signature verification happening in [the](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/ecdsa/ecdsa.go;l=111;drc=ea5d7cbc2644643331bd675b1ebdf0aaac7419f1) [ecdsa](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/ecdsa/ecdsa.go;l=214;drc=8c8429fe4113b399355c11203e60e6b37bc823ba) [call](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/ecdsa/ecdsa_noasm.go;l=15;drc=8c8429fe4113b399355c11203e60e6b37bc823ba) [stack](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/ecdsa/ecdsa.go;l=283;drc=8c8429fe4113b399355c11203e60e6b37bc823ba).

### Additional information

I believe that the golang stdlib ecdsa implementation is not _actually_ vulnerable to the signature fault attack as described above. In particular, that attack relies on the fact that many ecdsa implementations skirted the issue of bad random nonce generators by adopting deterministic nonces. It appears that the ecdsa implementation here [uses true randomness](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/ecdsa/ecdsa.go;l=240-261;drc=8c8429fe4113b399355c11203e60e6b37bc823ba), and is therefore not vulnerable. However I do believe that the implementation today _could_ silently produce faulty signatures, which we may want to avoid as both a correctness and a defense-in-depth measure.

**Comment by seankhliao** [comment]
cc @golang/security

**Comment by FiloSottile** [comment]
Faults can always cause incorrect outputs, for example by happening after the signature is generated and verified. This is a major problem if the fault leaks the private key, but crypto/ecdsa's implementation doesn't since the nonce is derived from both randomness, secrets, and the message, as you point out. I don't see why adding the complexity and performance cost of verifying the signature would help us.

A better argument could be made for Ed25519, though, which is deterministic.

**Comment by aarongable** [comment]
Yep, it [appears](https://research.kudelskisecurity.com/2017/10/04/defeating-eddsa-with-faults/) that Ed25519 is vulnerable to single fault attacks, and I don't see a verification step in the [Ed25519 signing implementation](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/ed25519/ed25519.go;l=151;drc=8c8429fe4113b399355c11203e60e6b37bc823ba).

Would you like me to repurpose this bug to track that, or file a separate one?
